src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -84,14 +85,72 @@ class SessionWrapperStack {
     SessionWrapper* m_back = nullptr;
 };
 
-struct ErrorTryAgainBackoffInfo {
-    void update(const ProtocolErrorInfo& info);
-    void reset();
-    std::chrono::milliseconds delay_interval();
+template <typename ErrorType, typename RandomEngine>
+struct ErrorBackoffState {
+    ErrorBackoffState() = default;
+    explicit ErrorBackoffState(ResumptionDelayInfo default_delay_info, RandomEngine& random_engine)
+        : default_delay_info(std::move(default_delay_info))
+        , m_random_engine(random_engine)
+    {
+    }
+
+    void update(ErrorType new_error, std::optional<ResumptionDelayInfo> new_delay_info)
+    {
+        if (triggering_error && *triggering_error == new_error) {
+            return;
+        }
+
+        delay_info = new_delay_info.value_or(default_delay_info);
+        cur_delay_interval = util::none;
+        triggering_error = new_error;
+    }
+
+    void reset()
+    {
+        triggering_error = util::none;
+        cur_delay_interval = util::none;
+        delay_info = default_delay_info;
+    }
 
+    std::chrono::milliseconds delay_interval()
+    {
+        if (!cur_delay_interval) {
+            cur_delay_interval = delay_info.resumption_delay_interval;
+            return jitter_value(*cur_delay_interval);
+        }
+        if (*cur_delay_interval >= delay_info.max_resumption_delay_interval) {
+            return jitter_value(delay_info.max_resumption_delay_interval);
+        }
+        auto safe_delay_interval = cur_delay_interval->count();
+        if (util::int_multiply_with_overflow_detect(safe_delay_interval,
+                                                    delay_info.resumption_delay_backoff_multiplier)) {
+            *cur_delay_interval = delay_info.max_resumption_delay_interval;
+        }
+        else {
+            *cur_delay_interval = std::chrono::milliseconds(safe_delay_interval);
```

Comment:
it looks like `safe_delay_interval` could go above `delay_info.max_resumption_delay_interval`. That doesn't really have any impact, but we could cap it at max_resumption_delay_interval:

`*cur_delay_interval = std::min(std::chrono::milliseconds(safe_delay_interval), delay_info.max_resumption_delay_interval);`

We could also update the check at line 121 to  `if ( ... == ... )`

Reply:
it's still a bit odd that we allow it once to go greater than `max_resumption_delay_interval` and then decrease it to max_resumption_delay_interval.

Reply:
Done.